python/ray/tune/tests/test_trial_scheduler_pbt.py

```diff
@@ -778,7 +783,162 @@ def test_config(
+    )
+    pb2_flat = create_pb2_scheduler(
+        metric="score",
+        mode="max",
+        hyperparam_bounds=flatten_dict(hyperparam_bounds, delimiter=""),
+    )
+
+    mock_runner = MagicMock()
+
+    trials_nested = [Trial("pb2_test", stub=True) for _ in range(3)]
+    trials_flat = [Trial("pb2_test", stub=True) for _ in range(3)]
+
+    np.random.seed(2023)
+
+    for trial_nested, trial_flat in zip(trials_nested, trials_flat):
+        pb2_nested.on_trial_add(mock_runner, trial_nested)
+        # Let PB2 generate the initial config randomly, then use the same
+        # initial values for the flattened version
+        flattened_init_config = flatten_dict(trial_nested.config, delimiter="")
+        trial_flat.config = flattened_init_config
+        pb2_flat.on_trial_add(mock_runner, trial_flat)
+
+    # Make sure that config suggestions are the same for each timestep
+    for t in range(1, 10):
+        for i, (trial_nested, trial_flat) in enumerate(zip(trials_nested, trials_flat)):
+            res = result(time=t, val=t * (i + 1) * 10)
+            save_trial_result(pb2_nested, trial_nested, t, res)
+            save_trial_result(pb2_flat, trial_flat, t, res)
+
+        new_config, _ = pb2_nested._get_new_config(trials_nested[0], trials_nested[-1])
+        new_config_flat, _ = pb2_flat._get_new_config(trials_flat[0], trials_flat[-1])
+
+        # Make sure the suggested config is still nested properly
+        assert list(new_config.keys()) == ["a", "b"]
+        assert list(new_config["b"].keys()) == ["c", "d"]
+        assert np.allclose(
+            list(flatten_dict(new_config, delimiter="").values()),
+            list(new_config_flat.values()),
+        )
+
+
+def test_pb2_missing_hyperparam_init():
+    """Test that PB2 fills in all missing hyperparameters (those that are not
+    specified in param_space)."""
+    hyperparam_bounds = {"a": [1.0, 2.0], "b": {"c": [2.0, 4.0], "d": [4.0, 10.0]}}
+    pb2 = create_pb2_scheduler(hyperparam_bounds=hyperparam_bounds)
+    mock_runner = MagicMock()
+
+    def validate_config(config, bounds):
+        for param, bound in bounds.items():
+            if isinstance(bound, dict):
+                validate_config(config[param], bound)
+            else:
+                low, high = bound
+                assert config[param] >= low and config[param] < high
+
+    trial = Trial("test_pb2", stub=True)
+    pb2.on_trial_add(mock_runner, trial)
+    validate_config(trial.config, hyperparam_bounds)
+
+    trial = Trial("test_pb2", stub=True, config={"b": {"c": 3.0}})
+    pb2.on_trial_add(mock_runner, trial)
+    validate_config(trial.config, hyperparam_bounds)
+    assert trial.config["b"]["c"] == 3.0
+
+
+def test_pb2_hyperparam_bounds_validation():
+    hyperparam_bounds = {"a": [1.0, 2.0], "b": {"c": [2.0, 4.0, 6.0]}}
+    with pytest.raises(ValueError):
+        create_pb2_scheduler(hyperparam_bounds=hyperparam_bounds)
+
+
+def test_pb2_custom_explore_fn():
```

Comment:
can we also check if the scheduler can be serialised if the function is set to eg. a lambda?